### PR TITLE
Support 32-bit platforms, take 2

### PR DESCRIPTION
**Pull request by shadaj**
Now rebased from master, with only the changes for 32-bit support.

This change is binary incompatible with Scala Native 0.3.x, since it introduces a new NIR type for results of `sizeOf` since we don't know the actual result type (`Int` or `Long`) inside the compiler plugin.

### CODE REVIEW

**Comment by shadaj** [comment]
Ping @densh, want to make sure this doesn't get too far out of date compared to master.

File: nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+object Platform {
+  def isMac(): Boolean                 = ExternPlatform.isMac()
+  def isWindows(): Boolean             = ExternPlatform.isWindows()
+  def windowsGetUserLang(): CString    = ExternPlatform.windowsGetUserLang()
+  def windowsGetUserCountry(): CString = ExternPlatform.windowsGetUserCountry()
+  def littleEndian(): Boolean          = ExternPlatform.littleEndian()
+  def setOSProps(addProp: CFunctionPtr2[CString, CString, Unit]): Unit =
+    ExternPlatform.setOSProps(addProp)
+
+  @inline final def cross3264[A, B](thirtyTwo: => A,
```
[the remark this answers is not in the document]
Comment (shadaj) [reply nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala]: The reason I defined the method like this was to handle different types across 32/64 bit platforms, such as returning an Int vs Long if a `CLong` is expected. Is there a way to allow this to be optimized without losing the type safety?

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala [hunk 2]
```diff
@@ -234,6 +234,7 @@ final class BinaryDeserializer(buffer: ByteBuffer) {
     case T.VoidType        => Type.Void
     case T.VarargType      => Type.Vararg
     case T.PtrType         => Type.Ptr
+    case T.SizeType        => sizeType
```
[the remark this answers is not in the document]
Comment (shadaj) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: I was considering that, but would that cause issues with link-time evaluation of math expressions where `Type.Size` would not be available for calculations?

File: tools/src/main/scala/scala/scalanative/codegen/MemoryLayout.scala [hunk 3]
```diff
@@ -32,52 +33,73 @@ object MemoryLayout {
...
+      case Type.Nothing | Type.Ptr | _: Type.RefKind =>
+        math.max((if (targetArchitecture.is32) 32 else 64) / WORD_SIZE, 1)
+      case _ => unsupported(s"sizeOf $ty")
+    }
+
+  def alignmentSize(ty: Type, targetArchitecture: TargetArchitecture) = {
+    if (targetArchitecture.is32) {
+      ty match {
+        case Type.Double => 4L
+        case Type.Long   => 4L
```
[the remark this answers is not in the document]
Comment (shadaj) [reply tools/src/main/scala/scala/scalanative/codegen/MemoryLayout.scala]: Yup, https://en.wikipedia.org/wiki/Data_structure_alignment#Typical_alignment_of_C_structs_on_x86!

File: tools/src/main/scala/scala/scalanative/build/TargetArchitecture.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ */
+sealed class TargetArchitecture(val bits: Bits,
+                                val isIntel: Boolean,
+                                val id: Int) {
+  val is32 = bits == ThirtyTwo
+}
+
+object TargetArchitecture {
+  case object i386   extends TargetArchitecture(ThirtyTwo, true, 1)
+  case object i686   extends TargetArchitecture(ThirtyTwo, true, 2)
```
[the remark this answers is not in the document]
Comment (shadaj) [reply tools/src/main/scala/scala/scalanative/build/TargetArchitecture.scala]: I'm hoping to also use `TargetArchitecture` for cross compilation at the LLVM level, so having each individual target will allow us to generate specific target triples.

File: unit-tests/src/test/scala/java/net/SocketSuite.scala [hunk 5]
```diff
@@ -45,13 +47,26 @@ object SocketSuite extends tests.Suite {
...
-  test("soTimeout") {
-    val s         = new Socket()
-    val prevValue = s.getSoTimeout
-    s.setSoTimeout(prevValue + 1000)
-    assertEquals(s.getSoTimeout, prevValue + 1000)
-    s.close()
-  }
+  Platform.cross3264(
+    {
+      // something is broken on i386
```
[the remark this answers is not in the document]
Comment (shadaj) [reply unit-tests/src/test/scala/java/net/SocketSuite.scala]: https://github.com/scala-native/scala-native/issues/1369

**Comment by shadaj** [comment]
@densh rebased and addressed feedback!

File: nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (densh) [reply nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala]: Indeed, there are benefits in terms of type safety. But from a pragmatic point of view, you can't have afford having the default implementation of equals and hashCode to allocate closures. A compromise where those kind of cases use a `if (isCross32) ... else ...` and the others (non-performance-sensitive) parts use the type safe version would be fine with me.

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (densh) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: Type.Size is special because it's effectively a type alias that we need to resolve at link time (rather than compile-time like in scalac). There is no machinery to dealias those kind of things atm, but I don't think putting platform-dependant logic into the deserializer is a good idea. Currently it's just 1-to-1 mapping without any logic whatsoever, it should rather stay this way.

File: unit-tests/src/test/scala/java/net/SocketSuite.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (densh) [reply unit-tests/src/test/scala/java/net/SocketSuite.scala]: 👍

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (densh) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: Actually if we had a way to dealias at link-time we wouldn't need separate artifacts for 32-bit and 64-bit platforms. We could have one NIR version of everything that's automagically become 32-bit or 64-bit based on the requested target architecture.

[the remark this answers is not in the document]
Comment (densh) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: But I think it's better left for next PR(s), it's likely not trivial and needs some design discussion as well.

File: tools/src/main/scala/scala/scalanative/build/TargetArchitecture.scala [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  case object x86_64 extends TargetArchitecture(SixtyFour, true, 3)
+  case object arm    extends TargetArchitecture(ThirtyTwo, false, 4)
```
[the remark this answers is not in the document]
Comment (densh) [reply tools/src/main/scala/scala/scalanative/build/TargetArchitecture.scala]: We should have failing tests disabled for that platform and file corresponding issues to keep the progress on getting to 100% coverage.

File: tools/src/main/scala/scala/scalanative/codegen/MemoryLayout.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (densh) [reply tools/src/main/scala/scala/scalanative/codegen/MemoryLayout.scala]: 👍

File: tools/src/main/scala/scala/scalanative/build/TargetArchitecture.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (densh) [reply tools/src/main/scala/scala/scalanative/build/TargetArchitecture.scala]: Sounds good.

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (shadaj) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: I ended up being able to eliminate `Type.Size` altogether, actually, by figuring out what the `Word` type is in the compiler plugin. Especially since native libraries sometimes use types that differ across 32/64bit systems, I think link-time switching won't work since it wouldn't be possible to express these differing APIs.

File: nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (shadaj) [reply nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala]: That makes sense. Is there a reason there would still be closures allocated even after `cross3264` is inlined?

**Comment by shadaj** [comment]
Okay something is seriously goofed. I can't seem to be able to reproduce the segfault on my computer with the same image. But it's happening consistently on Travis, looks like something with running out of memory since it happens on random suites. @densh could you see if you get the segfault when running the tests?

Commands to run locally:
```bash
export TARGET_DOCKER_PLATFORM=i386; docker build -t scala-native-testing:linux-$TARGET_DOCKER_PLATFORM --build-arg TARGET_DOCKER_PLATFORM=$TARGET_DOCKER_PLATFORM --build-arg HOST_ARCHITECTURE=amd64 --cpuset-cpus=0 ci-docker

export TARGET_DOCKER_PLATFORM=i386; export SCALANATIVE_GC=boehm; export SBT_VERSION=0.13.16; export TEST_COMMAND="tests_x32/test"; ./ci-docker/run-test.sh
```

**Comment by shadaj** [comment]
Nevermind, adjusting how Boehm is built was enough to fix it!

**Comment by shadaj** [comment]
The Boehm goofing has returned 😕. I'll continue to work on debugging this but in terms of the changes to Scala Native itself ready for another review @densh!

**Comment by shadaj** [comment]
Everything is green again! Ready for another review @densh!

File: ci-docker/Dockerfile [hunk 7]
```diff
@@ -1,13 +1,15 @@
 ARG TARGET_DOCKER_PLATFORM
 
-FROM ubuntu:18.04
+FROM ubuntu:18.10
```
Comment (densh) [inline ci-docker/Dockerfile]: Any reason to use newer Ubuntu image? I think we should stick to 18.04 as it's latest LTS release atm.

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 8]
```diff
@@ -173,7 +177,7 @@ private[net] class PlainSocketImpl extends SocketImpl {
   override def connect(address: SocketAddress, timeout: Int): Unit = {
     val inetAddr = address.asInstanceOf[InetSocketAddress]
     val hints    = stackalloc[addrinfo]
-    val ret      = stackalloc[Ptr[addrinfo]]
+    val ret      = stdlib.malloc(sizeof[Ptr[Byte]]).cast[Ptr[Ptr[addrinfo]]] //stackalloc[Ptr[addrinfo]]
```
Comment (densh) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: Stackalloc and malloc are not 1-to-1 equivalents. malloc need to be cleaned up manually otherwise this line introduces a memory leak.

File: nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (densh) [reply nativelib/_x32/src/main/scala/scala/scalanative/runtime/Platform.scala]: LLVM's IR is around C level's of abstraction (structs, arrays and pointers). So even after inlining of the `cross3264` it would still keep the allocation and virtual dispatch as it doesn't really know how to deal with those.

File: tools/src/main/scala/scala/scalanative/linker/ClassPath.scala [hunk 9]
```diff
@@ -20,14 +20,16 @@ sealed trait ClassPath {
 object ClassPath {
 
   /** Create classpath based on the directory. */
-  def apply(directory: Path): ClassPath =
-    new Impl(VirtualDirectory.local(directory))
+  def apply(directory: Path, config: build.Config): ClassPath =
+    new Impl(VirtualDirectory.local(directory), config)
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/ClassPath.scala]: It looks like config is not used any more. It was probably added previously to support previous version of sizeof that desugared in binary serializer.

File: tools/src/main/scala/scala/scalanative/linker/Reach.scala [hunk 10]
```diff
@@ -544,8 +544,9 @@ class Reach(config: build.Config, entries: Seq[Global], loader: ClassLoader) {
       reachVal(v)
     case Op.Copy(v) =>
       reachVal(v)
-    case Op.Sizeof(ty) =>
+    case Op.Sizeof(ty, retType) =>
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/Reach.scala]: We usually use `ty` as a shortcut for `type` and `retty` name as a shortcut for `return type`.

File: tools/src/main/scala/scala/scalanative/codegen/Metadata.scala [hunk 11]
```diff
@@ -6,7 +6,11 @@ import scalanative.nir._
 import scalanative.sema._
 import scalanative.linker.{Trait, Struct, Class}
 
-class Metadata(val linked: linker.Result, proxies: Seq[Defn]) {
+import scala.scalanative.build.TargetArchitecture
+
+class Metadata(val linked: linker.Result,
+               proxies: Seq[Defn],
+               val targetArchitecture: TargetArchitecture) {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/codegen/Metadata.scala]: It looks like `targetArchitecture` is not used here.

File: tools/src/main/scala/scala/scalanative/build/Discover.scala [hunk 12]
```diff
@@ -90,6 +90,25 @@ object Discover {
...
+   * @param triple the LLVM target triple, see Discover.targetTriple
+   * @return the detected target architecture
+   */
+  def targetArchitecture(triple: String): TargetArchitecture = {
+    import TargetArchitecture._
+    triple.split('-').head match {
+      case "x86_64"  => x86_64
+      case "i386"    => i386
+      case "i686"    => i686
+      case "armv7l"  => arm
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/build/Discover.scala]: If we're using LLVM target architecture names, it's probably best to keep them as-is (i.e. `armv7l` on the right here). Otherwise there is a chance of confusing name clashes.

File: tools/src/main/scala/scala/scalanative/build/Config.scala [hunk 13]
```diff
@@ -68,6 +71,9 @@ sealed trait Config {
   /** Create a new config with given compilation options. */
   def withCompileOptions(value: Seq[String]): Config
 
+  /** Create a new config with given target architecture. */
+  def withTargetArchitecture(value: TargetArchitecture): Config
+
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/build/Config.scala]: Do we really need to have both `target triple` and `target architecture` as two separate concepts in the public API? It seems like `TargetArchitecture` is derivative from the triple and we can only have triple without changing the build API.

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala [hunk 14]
```diff
@@ -166,7 +182,8 @@ object ScalaNativePluginInternal {
     scalaNativeConfigSettings ++ Seq(
       test := (test in NativeTest).value,
       testOnly := (testOnly in NativeTest).evaluated,
-      testQuick := (testQuick in NativeTest).evaluated
+      testQuick := (testQuick in NativeTest).evaluated,
+      nativeLink := (nativeLink in NativeTest).value
```
Comment (densh) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: Why is this addition necessary?

File: project/build.sbt [hunk 15]
```diff
@@ -17,7 +17,8 @@ libraryDependencies ++= Seq(
   "org.eclipse.jgit" % "org.eclipse.jgit.pgm" % "3.2.0.201312181205-r"
 )
 
-addSbtPlugin("org.portable-scala" % "sbt-platform-deps" % "1.0.0-M2")
+addSbtPlugin("org.portable-scala" % "sbt-platform-deps" % "1.0.0")
+addSbtPlugin("org.portable-scala" % "sbt-crossproject"  % "0.6.0")
```
Comment (densh) [inline project/build.sbt]: It's a bit unfortunate that we have to depend on  cross project here, but I don't have an easy fix for it.

File: javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala [hunk 16]
```diff
@@ -25,10 +26,14 @@ final class NativePosixFileAttributeView(path: Path, options: Array[LinkOption])
 
     val buf = alloc[utime.utimbuf]
     !(buf._1) =
-      if (lastAccessTime != null) lastAccessTime.to(TimeUnit.SECONDS)
+      if (lastAccessTime != null)
+        Platform.cross3264(lastAccessTime.to(TimeUnit.SECONDS).toInt,
+                           lastAccessTime.to(TimeUnit.SECONDS))
```
Comment (densh) [inline javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: We put curlies around multi-line control-flow expressions.

**Comment by densh** [comment]
Btw, what the status of GCs on 32-bit?

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala (hunk 14 above)
Comment (shadaj) [reply sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: This makes it possible to run `test:nativeLink` for quickly testing LLVM compilation without actually running. It's especially useful when testing on macOS where running the container can get pretty slow.

File: tools/src/main/scala/scala/scalanative/codegen/Metadata.scala (hunk 11 above)
Comment (shadaj) [reply tools/src/main/scala/scala/scalanative/codegen/Metadata.scala]: It's used when calculating field layouts, as `meta.targetArchitecture`.

**Comment by densh** [comment]
Just opened a ticket #1376. I think we can get by without using cross projects. This PR is perfectly fine the way it is now, we can merge and then change to a different scheme some time before 0.4.0 release.

**Comment by shadaj** [comment]
How would you imagine static typing working? I know some libraries end up having quite different APIs when you are running on 32bit or 64bit. Would all `Word` types just default to `Int` so both platforms work?

**Comment by densh** [comment]
@shadaj Rough sketch with how it might work in terms of types:
```scala
// Get replaced with either i32 or i64 based on the target architecture at link-time.
@struct class Word private ()

implicit class RichInt(val value: Int) {
   // Explicit conversion from scala's Int to word.
   def toWord: Word = intrinsic
}

@extern object libfoo {
  def foo(word: Word): Unit = extern
}

libfoo.foo(42.toWord)
```
So Word is not an alias any more but a magical struct that gets replaced with either i32 or i64 based on current architecture. C numeric types would work similarly.

**Comment by shadaj** [comment]
Re GCs: right now only Boehm is supported. I think the general direction going forward for PRs after this one:
- Switch to link-time selection of word types
- Enable Immix on 32bit platforms
- Enable testing on ARM32/ARM64

**Comment by shadaj** [comment]
(╯°□°）╯︵ ┻━┻

I really have no clue what's going on with the sporadic segfaults happening on Travis. On my local computer with a freshly built image, all the unit tests ran fine with no errors. Could you try it locally @densh?

> ```shell
> export TARGET_DOCKER_PLATFORM=i386; docker build -t scala-native-testing:linux-$TARGET_DOCKER_PLATFORM --build-arg TARGET_DOCKER_PLATFORM=$TARGET_DOCKER_PLATFORM --build-arg HOST_ARCHITECTURE=amd64 --cpuset-cpus=0 ci-docker
> 
> export TARGET_DOCKER_PLATFORM=i386; export SCALANATIVE_GC=boehm; export SBT_VERSION=0.13.16; export TEST_COMMAND="tests_x32/test"; ./ci-docker/run-test.sh
> ```

**Comment by shadaj** [comment]
Did a bit of poking around, and looks like at least when running locally the tests segfault on every other run. Any clue why this might be happening?

**Comment by shadaj** [comment]
Running the unit tests separately helped a lot, I was able to run with GDB and found that there is a segfault occurring when the stride element of array headers is being set. Since that value is only used for Immix, which does not work for 32bit platforms, I've just changed that logic to not set the value if on a 32bit platform.

**Comment by Sciss** [comment]
@shadaj sorry to hijack the thread - about your example compiling SN to wasm; how difficult is it to reproduce? do you have a SN fork with this PR applied? It would be so great to have a tutorial or blog post about getting [this](https://twitter.com/ShadajL/status/1010730085303795713) done. Thanks!

**Comment by shadaj** [comment]
Hi @sciss, I won't have access to my laptop through next week, but I'll share my example project as soon as I get back!

**Comment by gurinderu** [comment]
@shadaj great work. thanks.  But I still a little bit confused about the structure of os dependent code. 
It is important for javalib and threads especially.

**Comment by shadaj** [comment]
@Sciss just pushed up the demo to https://github.com/shadaj/scala-native-wasm!

**Comment by shadaj** [comment]
Closing in favor of #1571

**Comment by no-identd** [comment]
Hi. The README.MD of https://github.com/shadaj/scala-native-wasm (the implicitly first result when googling "Scala 3" "wasm") still points here. If you found your way here not in response to a github notification or similar, you'll probably want to take a look at the following besides just #1571:
#2502
#2574
#2620

But most important, and foremost, of all:

https://mobile.twitter.com/ShadajL/status/1548020571597811719
